### PR TITLE
Add markdownlint as a pytool plugin [Rebase & FF]

### PR DESCRIPTION
**Pull request by kenlautner**
## Description

Cherry-pick commits that add markdownlint as a pytool plugin as well as markdown fixes in several MD files.
Commits from 202311 included here:

df8805cf8f
6d509b255e
72f4ac0dfc
40da3f2b31
984dee96cf
c149b3f3e3
aa5ae2af5d
26eb685974
26e38dc194

- [ ] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [ ] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

N/A

## Integration Instructions

N/A

### CODE REVIEW

File: .pytool/Plugin/SpellCheck/cspell.base.yaml [hunk 1]
```diff
@@ -291,8 +291,6 @@
         "unrecovered",
         "cmocka",
         "unenrolling",
-        "unconfigure",
-        "Loongson",
-        "LOONGARCH"
+        "unconfigure"
```
Comment (os-d) [inline .pytool/Plugin/SpellCheck/cspell.base.yaml]: We should drop this as it is already there and this removes other spell check words.

File: UnitTestFrameworkPkg/ReadMe.md [hunk 2]
```diff
@@ -1301,20 +1313,6 @@ leave those details to be explained in the gMock documentation.
...
-When using the EDK2 Pytools for CI testing, the host-based unit tests will be built and run on any build that includes
-the `NOOPT` build target.
-
-If you are trying to iterate on a single test, a convenient pattern is to build only that test module. For example,
-the following command will build only the SafeIntLib host-based test from the MdePkg...
-
-```bash
-stuart_ci_build -c .pytool/CISettings.py TOOL_CHAIN_TAG=VS2017 -p MdePkg -t NOOPT BUILDMODULE=MdePkg/Test/UnitTest/Library/BaseSafeIntLib/TestBaseSafeIntLib.inf
-```
-
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: I don't think this should be dropped?

File: UnitTestFrameworkPkg/ReadMe.md [hunk 3]
```diff
@@ -1096,6 +1092,22 @@ int main(int argc, char* argv[]) {
...
+can be used to add them into the file indirectly.
+
+The reason for this is due to EDK II taking the host application INF file and
+first compiling all of its source files into a static library. This static
+library is then linked into the final host application. The problem with this
+method is that only the tests in the object file containing the `main()`
+function are linked into the final host application. This is because the other
+tests are contained in their own object files within the static library and
+they have no symbols in them that the final host application depends on, so
+those object files are not linked into the final host application.
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: This would appear not to be related to markdown linting

File: UnitTestFrameworkPkg/ReadMe.md [hunk 4]
```diff
@@ -1364,6 +1362,10 @@ After that, the following commands will set up the build and run the host-based
 # stuart_setup -c ./.pytool/CISettings.py TOOL_CHAIN_TAG=<GCC5, VS2019, etc.>
 stuart_setup -c ./.pytool/CISettings.py TOOL_CHAIN_TAG=VS2019
 
+# Mu specific step to clone mu repos required for ci check
+# stuart_ci_setup -c ./.pytool/CISettings.py TOOL_CHAIN_TAG=<GCC5, VS2019, etc.>
+stuart_ci_setup -c ./.pytool/CISettings.py TOOL_CHAIN_TAG=VS2019
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: Similarly, not related to markdown linting

File: UnitTestFrameworkPkg/ReadMe.md [hunk 5]
```diff
@@ -1557,7 +1562,7 @@ Code/Test                                   | Location
...
+Non-Host-Based (PEI/DXE/SMM/UefiShell) Tests for a Functionality or Feature   | Similar to Host-Based, if the feature is in one package, should be located in the `*Pkg/Test/[UefiShell/Dxe/Smm/Pei]Test` directory.<br/><br/>If the feature spans multiple packages, it's location should be determined by the package owners related to the feature.<br/><br/>USAGE EXAMPLES<br/>PEI Example: MP_SERVICE_PPI. Or check MTRR configuration in a notification function.<br/> SMM Example: a test in a protocol callback function. (It is different with the solution that SmmAgent+ShellApp)<br/>DXE Example: a test in a UEFI event call back to check SPI/SMRAM status. <br/> Shell Example: the SMM handler audit test has a shell-based app that interacts with an SMM handler to get information. The SMM paging audit test gathers information about both DXE and SMM. And the SMM paging functional test actually forces errors into SMM via a DXE driver.
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: I think also not related to markdown linting. I suspect the linting occurred after some other changes to this file, I believe these are being taken care of in another PR.

File: UnitTestFrameworkPkg/ReadMe.md [hunk 6]
```diff
@@ -1630,14 +1635,15 @@ Non-Host-Based (PEI/DXE/SMM/Shell) Tests for a Functionality or Feature   | Simi
...
+the following, please make sure they live in the correct place.
 
 Code/Test                                   | Location
 ---------                                   | --------
-Host-Based Library Implementations                 | Host-Based Implementations of common libraries (eg. MemoryAllocationLibHost) should live in the same package that declares the library interface in its .DEC file in the `*Pkg/HostLibrary` directory. Should have 'Host' in the name.
-Host-Based Mocks and Stubs  | Mock and Stub libraries should live in the `UefiTestFrameworkPkg/StubLibrary` with either 'Mock' or 'Stub' in the library name.
+Host-Based Library Implementations                 | Host-Based Implementations of common libraries (eg. MemoryAllocationLibHost) should live in the same package that declares the library interface in its .DEC file in the `*Pkg/Test/Library` directory. Should have 'Host' in the name.
+Host-Based Mocks and Stubs  | Mock and Stub libraries that require test infrastructure should live in the `UefiTestFrameworkPkg/Library` with either 'Mock' or 'Stub' in the library name.
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: All this between the last comment also seems unrelated to linting.

File: MdeModulePkg/Library/VariablePolicyLib/ReadMe.md [hunk 7]
```diff
@@ -4,8 +4,6 @@ version:    1.0
 copyright:  Copyright (c) Microsoft Corporation.
 ---
 
-# UEFI Variable Policy
```
Comment (os-d) [inline MdeModulePkg/Library/VariablePolicyLib/ReadMe.md:7]: This should stay?

Comment (kenlautner) [reply MdeModulePkg/Library/VariablePolicyLib/ReadMe.md:7]: One of the higher level headings needed to be dropped.  I thought the Uefi Variable policy one made more sense to get rid of.

File: .pytool/Plugin/SpellCheck/cspell.base.yaml (hunk 1 above)
Comment (kenlautner) [reply .pytool/Plugin/SpellCheck/cspell.base.yaml]: Fixed.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 2 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: Readded.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 3 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: You're correct.  There was a PR that included some markdown fixes and updated the document.  I removed the documentation updates that aren't related to markdown fixes.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 4 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: You're correct.  There was a PR that included some markdown fixes and updated the document.  I removed the documentation updates that aren't related to markdown fixes.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 5 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: You're correct.  There was a PR that included some markdown fixes and updated the document.  I removed the documentation updates that aren't related to markdown fixes.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 6 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: You're correct.  There was a PR that included some markdown fixes and updated the document.  I removed the documentation updates that aren't related to markdown fixes.

File: UnitTestFrameworkPkg/ReadMe.md [hunk 8]
```diff
@@ -66,11 +66,6 @@ reviewed. The paths to the SecureBootVariableLib unit tests are:
 | Type-Parameterized Tests    |    NO     |    YES     |
 | Timeout Support             |    NO     |    YES     |
 | Mocking Support             |   Cmocka  |   gMock    |
-| JUNIT XML Reports           |    YES    |    YES     |
-| Execute subset of tests     |    NO     |    YES     |
-| VS Code Extensions          |    NO     |    YES     |
-
-## Framework Libraries
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: I believe we want to keep this?

File: UnitTestFrameworkPkg/ReadMe.md [hunk 9]
```diff
@@ -1611,7 +1611,7 @@ Non-Host-Based (PEI/DXE/SMM/Shell) Tests for a Functionality or Feature   | Simi
 
       GeneralPurposeLib/              # Host-Based Test for any implementation of GeneralPurposeLib
         GeneralPurposeLibTest.c
-        GeneralPurposeLibHostUnitTest.inf
+        GeneralPurposeLibHostUnitTest
```
Comment (os-d) [inline UnitTestFrameworkPkg/ReadMe.md]: We would seem to want to keep this, too?

**Review by os-d — APPROVED** [review]

File: .azurepipelines/templates/markdownlint-check-prereq-steps.yml
Comment (makubacki) [inline .azurepipelines/templates/markdownlint-check-prereq-steps.yml:1]: It looks like the last 4 commits (fixes) can be squashed with a co-author.

File: .azurepipelines/templates/pr-gate-steps.yml
Comment (makubacki) [inline .azurepipelines/templates/pr-gate-steps.yml:1]: This file addition shouldn't be in markdownlint related changes.

File: BaseTools/Scripts/PackageDocumentTools/Readme.md
Comment (makubacki) [inline BaseTools/Scripts/PackageDocumentTools/Readme.md:1]: This file has unrelated changes.

**Review by makubacki — APPROVED** [review]

File: .azurepipelines/templates/markdownlint-check-prereq-steps.yml
Comment (kenlautner) [reply .azurepipelines/templates/markdownlint-check-prereq-steps.yml:1]: Merged.

File: .azurepipelines/templates/pr-gate-steps.yml
Comment (kenlautner) [reply .azurepipelines/templates/pr-gate-steps.yml:1]: I removed the .azurepipelines files as they we're a relic of the previous CI system.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 8 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: That's correct.  I readded the lines.

File: UnitTestFrameworkPkg/ReadMe.md (hunk 9 above)
Comment (kenlautner) [reply UnitTestFrameworkPkg/ReadMe.md]: Fixed.

**Pull request merged**